diff --git a/src/Actions.c b/src/Actions.c
index abb61ef..1048249 100644
--- a/src/Actions.c
+++ b/src/Actions.c
@@ -151,13 +151,15 @@ static Cardinal num_variable_list;
 Bool
 XawParseBoolean(Widget w, String param, XEvent *event, Bool *succeed)
 {
+  char *tmp = (char *)param;
+  int value;
+  double dd = strtod(param, &tmp);
+
   if (!param)
     return (False);
   else
     {
-      char *tmp = (char *)param;
-      double dd = strtod(param, &tmp);
-      int value = (int) dd;
+      value = (int) dd;
 
       if (*tmp == '\0')
         return (value);
diff --git a/src/AsciiSink.c b/src/AsciiSink.c
index 1736a76..cda6f37 100644
--- a/src/AsciiSink.c
+++ b/src/AsciiSink.c
@@ -1633,7 +1633,7 @@ GetGC(AsciiSinkObject sink)
 			  GCForeground | GCBackground);
     XGCValues values = {
 	/* XXX We dont want to share a gc that will change the clip-mask */
-	.clip_x_origin = (int)(long)sink,
+	.clip_x_origin = (int)(uintptr_t)sink,
 	.clip_mask = None,
 	.font = sink->ascii_sink.font->fid,
 	.graphics_exposures = False,
diff --git a/src/AsciiSrc.c b/src/AsciiSrc.c
index a3703ae..e2e8881 100644
--- a/src/AsciiSrc.c
+++ b/src/AsciiSrc.c
@@ -1488,7 +1488,7 @@ InitStringOrFile(AsciiSrcObject src, Bool newString)
 	int fd = open(src->ascii_src.string, (int)open_mode, 0666);
 
 	if (fd != -1) {
-	    FILE *file = fdopen(fd, fdopen_mode);
+	    FILE *file = _fdopen(fd, fdopen_mode);
 
 	    if (file != NULL) {
 		(void)fseek(file, 0, SEEK_END);
@@ -1788,7 +1788,7 @@ GetDefaultPieceSize(Widget w _X_UNUSED, int offset _X_UNUSED, XrmValue *value)
     static XPointer pagesize;
 
     if (pagesize == NULL) {
-	pagesize = (XPointer)((long)_XawGetPageSize());
+	pagesize = (XPointer)((uintptr_t)_XawGetPageSize());
 	if (pagesize < (XPointer)BUFSIZ)
 	    pagesize = (XPointer)BUFSIZ;
     }
diff --git a/src/Command.c b/src/Command.c
index a2b298a..a865a46 100644
--- a/src/Command.c
+++ b/src/Command.c
@@ -570,10 +570,10 @@ XawCommandGetValuesHook(Widget w, ArgList args, Cardinal *num_args)
     for (i = 0; i < *num_args; i++) {
 	if (STR_EQUAL(args[i].name, XtNforeground))
 	    *((String*)args[i].value) = cbw->command.set ?
-		(String)cbw->core.background_pixel : (String)cbw->label.foreground;
+		(String)(uintptr_t)cbw->core.background_pixel : (String)(uintptr_t)cbw->label.foreground;
 	else if (STR_EQUAL(args[i].name, XtNbackground))
 	    *((String*)args[i].value) = cbw->command.set ?
-		(String)cbw->label.foreground : (String)cbw->core.background_pixel;
+		(String)(uintptr_t)cbw->label.foreground : (String)(uintptr_t)cbw->core.background_pixel;
     }
 }
 
diff --git a/src/DisplayList.c b/src/DisplayList.c
index 457485c..8f2e325 100644
--- a/src/DisplayList.c
+++ b/src/DisplayList.c
@@ -789,7 +789,7 @@ DlForeground(Widget w, XtPointer args, XtPointer data,
 	     XEvent *event _X_UNUSED, Region region _X_UNUSED)
 {
   XawXlibData *xdata = (XawXlibData *)data;
-  Pixel foreground = (Pixel)args;
+  Pixel foreground = (Pixel)(uintptr_t)args;
 
   if (xdata->values.foreground != foreground)
     {
@@ -805,7 +805,7 @@ DlBackground(Widget w, XtPointer args, XtPointer data,
 	     XEvent *event _X_UNUSED, Region region _X_UNUSED)
 {
   XawXlibData *xdata = (XawXlibData *)data;
-  Pixel background = (Pixel)args;
+  Pixel background = (Pixel)(uintptr_t)args;
 
   if (xdata->values.background != background)
     {
@@ -912,7 +912,7 @@ DlLineWidth(Widget w, XtPointer args, XtPointer data,
 	    XEvent *event _X_UNUSED, Region region _X_UNUSED)
 {
   XawXlibData *xdata = (XawXlibData *)data;
-  unsigned line_width = (unsigned)(unsigned long)args;
+  unsigned line_width = (unsigned)(uintptr_t)args;
 
   if ((unsigned)xdata->values.line_width != line_width)
     {
@@ -987,7 +987,7 @@ DlArcMode(Widget w, XtPointer args, XtPointer data,
 	  XEvent *event _X_UNUSED, Region region _X_UNUSED)
 {
   XawXlibData *xdata = (XawXlibData *)data;
-  int arc_mode  = (int)(long)args;
+  int arc_mode  = (int)(uintptr_t)args;
 
   if (xdata->values.arc_mode != arc_mode)
     {
@@ -1003,7 +1003,7 @@ DlCoordMode(Widget w _X_UNUSED, XtPointer args, XtPointer data,
 	    XEvent *event _X_UNUSED, Region region _X_UNUSED)
 {
   XawXlibData *xdata = (XawXlibData *)data;
-  int mode  = (int)(long)args;
+  int mode  = (int)(uintptr_t)args;
 
   xdata->mode = mode;
 }
@@ -1014,7 +1014,7 @@ DlShapeMode(Widget w _X_UNUSED, XtPointer args, XtPointer data,
 	    XEvent *event _X_UNUSED, Region region _X_UNUSED)
 {
   XawXlibData *xdata = (XawXlibData *)data;
-  int shape  = (int)(long)args;
+  int shape  = (int)(uintptr_t)args;
 
   xdata->shape = shape;
 }
@@ -1025,7 +1025,7 @@ DlLineStyle(Widget w, XtPointer args, XtPointer data,
 	    XEvent *event _X_UNUSED, Region region _X_UNUSED)
 {
   XawXlibData *xdata = (XawXlibData *)data;
-  int line_style = (int)(long)args;
+  int line_style = (int)(uintptr_t)args;
 
   if (xdata->values.line_style != line_style)
     {
@@ -1041,7 +1041,7 @@ DlCapStyle(Widget w, XtPointer args, XtPointer data,
 	   XEvent *event _X_UNUSED, Region region _X_UNUSED)
 {
   XawXlibData *xdata = (XawXlibData *)data;
-  int cap_style = (int)(long)args;
+  int cap_style = (int)(uintptr_t)args;
 
   if (xdata->values.cap_style != cap_style)
     {
@@ -1057,7 +1057,7 @@ DlJoinStyle(Widget w, XtPointer args, XtPointer data,
 	    XEvent *event _X_UNUSED, Region region _X_UNUSED)
 {
   XawXlibData *xdata = (XawXlibData *)data;
-  int join_style = (int)(long)args;
+  int join_style = (int)(uintptr_t)args;
 
   if (xdata->values.join_style != join_style)
     {
@@ -1073,7 +1073,7 @@ DlFillStyle(Widget w, XtPointer args, XtPointer data,
 	    XEvent *event _X_UNUSED, Region region _X_UNUSED)
 {
   XawXlibData *xdata = (XawXlibData *)data;
-  int fill_style = (int)(long)args;
+  int fill_style = (int)(uintptr_t)args;
 
   if (xdata->values.fill_style != fill_style)
     {
@@ -1089,7 +1089,7 @@ DlFillRule(Widget w, XtPointer args, XtPointer data,
 	   XEvent *event _X_UNUSED, Region region _X_UNUSED)
 {
   XawXlibData *xdata = (XawXlibData *)data;
-  int fill_rule = (int)(long)args;
+  int fill_rule = (int)(uintptr_t)args;
 
   if (xdata->values.fill_rule != fill_rule)
     {
@@ -1144,7 +1144,7 @@ DlFunction(Widget w, XtPointer args, XtPointer data,
 	   XEvent *event _X_UNUSED, Region region _X_UNUSED)
 {
   XawXlibData *xdata = (XawXlibData *)data;
-  int function = (int)(long)args;
+  int function = (int)(uintptr_t)args;
 
   if (function != xdata->values.function)
     {
@@ -1160,7 +1160,7 @@ DlPlaneMask(Widget w, XtPointer args, XtPointer data,
 	    XEvent *event _X_UNUSED, Region region _X_UNUSED)
 {
   XawXlibData *xdata = (XawXlibData *)data;
-  unsigned long plane_mask = (unsigned long)args;
+  unsigned long plane_mask = (unsigned long)(uintptr_t)args;
 
   if (xdata->values.plane_mask != plane_mask)
     {
@@ -1227,7 +1227,7 @@ DlFont(Widget w, XtPointer args, XtPointer data,
        XEvent *event _X_UNUSED, Region region _X_UNUSED)
 {
   XawXlibData *xdata = (XawXlibData *)data;
-  Font font = (Font)args;
+  Font font = (Font)(uintptr_t)args;
 
   if (xdata->values.font != font)
     {
@@ -1259,7 +1259,7 @@ DlSubwindowMode(Widget w, XtPointer args, XtPointer data,
 		XEvent *event _X_UNUSED, Region region _X_UNUSED)
 {
   XawXlibData *xdata = (XawXlibData *)data;
-  int subwindow_mode = (int)(long)args;
+  int subwindow_mode = (int)(uintptr_t)args;
 
   if (xdata->values.subwindow_mode != subwindow_mode)
     {
@@ -1275,7 +1275,7 @@ DlExposures(Widget w, XtPointer args, XtPointer data,
 	    XEvent *event _X_UNUSED, Region region _X_UNUSED)
 {
   XawXlibData *xdata = (XawXlibData *)data;
-  Bool graphics_exposures = (Bool)(long)args;
+  Bool graphics_exposures = (Bool)(uintptr_t)args;
 
   if (xdata->values.graphics_exposures != graphics_exposures)
     {
@@ -1737,7 +1737,7 @@ _Xaw_Xlib_ArgsInitProc(String proc_name, String *params, Cardinal *num_params,
 	if (*num_params == 1 &&
 	    XAllocNamedColor(DisplayOfScreen(screen), colormap,
 			     params[0], &xcolor, &xcolor))
-	  retval = (void *)xcolor.pixel;
+	  retval = (void *)(uintptr_t)xcolor.pixel;
       } break;
     case MASK:
     case UMASK:
@@ -1748,7 +1748,7 @@ _Xaw_Xlib_ArgsInitProc(String proc_name, String *params, Cardinal *num_params,
       if (*num_params == 1)
 	{
 	  long x = read_int((char *)params[0], NULL);
-	  retval = (void *)x;
+	  retval = (void *)(uintptr_t)x;
 	}
       break;
     case ARCMODE:
@@ -1896,7 +1896,7 @@ _Xaw_Xlib_ArgsInitProc(String proc_name, String *params, Cardinal *num_params,
       if (*num_params == 1)
 	{
 	  long x = read_int((char *)params[0], NULL);
-	  retval = (void *)x;
+	  retval = (void *)(uintptr_t)x;
 	}
       break;
     case DSTRING:
@@ -1917,7 +1917,7 @@ _Xaw_Xlib_ArgsInitProc(String proc_name, String *params, Cardinal *num_params,
       if (*num_params == 1)
 	{
 	  Font x = XLoadFont(DisplayOfScreen(screen), params[0]);
-	  retval = (void *)x;
+	  retval = (void *)(uintptr_t)x;
 	}
       break;
     case DASHES:
@@ -1948,7 +1948,7 @@ _Xaw_Xlib_ArgsInitProc(String proc_name, String *params, Cardinal *num_params,
 	  if (isdigit((unsigned char)params[0][0]) || params[0][0] == '+' || params[0][0] == '-')
 	    {
 	      long x = read_int((char *)params[0], NULL);
-	      retval = (void *)x;
+	      retval = (void *)(uintptr_t)x;
 	    }
 	  else if (XmuCompareISOLatin1(params[0], "true") == 0 ||
 	    XmuCompareISOLatin1(params[0], "on") == 0)
diff --git a/src/MultiSink.c b/src/MultiSink.c
index 6bb2465..13227eb 100644
--- a/src/MultiSink.c
+++ b/src/MultiSink.c
@@ -659,7 +659,7 @@ GetGC(MultiSinkObject sink)
 			  GCForeground | GCBackground);
     XGCValues values = {
 	/* XXX We dont want to share a gc that will change the clip-mask */
-	.clip_x_origin = (int)(long)sink,
+	.clip_x_origin = (int)(uintptr_t)sink,
 	.clip_mask = None,
 	.graphics_exposures = False,
 
diff --git a/src/MultiSrc.c b/src/MultiSrc.c
index 5d23396..a97ea64 100644
--- a/src/MultiSrc.c
+++ b/src/MultiSrc.c
@@ -1256,7 +1256,7 @@ InitStringOrFile(MultiSrcObject src, Bool newString)
 
     if (!src->multi_src.is_tempfile) {
 	if ((fd = open((char *)src->multi_src.string, (int)open_mode, 0666)) != -1) {
-	    if ((file = fdopen(fd, fdopen_mode)) != NULL) {
+	    if ((file = _fdopen(fd, fdopen_mode)) != NULL) {
 		(void)fseek(file, 0, SEEK_END);
 		src->multi_src.length = (XawTextPosition)ftell(file);
 		return(file);
@@ -1611,7 +1611,7 @@ GetDefaultPieceSize(Widget w _X_UNUSED, int offset _X_UNUSED, XrmValue *value)
     static XPointer pagesize;
 
     if (pagesize == 0) {
-	pagesize = (XPointer)((long)_XawGetPageSize());
+	pagesize = (XPointer)((uintptr_t)_XawGetPageSize());
 	if (pagesize < (XPointer)BUFSIZ)
 	    pagesize = (XPointer)BUFSIZ;
     }
diff --git a/src/Pixmap.c b/src/Pixmap.c
index 0083b8c..708250c 100644
--- a/src/Pixmap.c
+++ b/src/Pixmap.c
@@ -47,7 +47,7 @@
  * Types
  */
 typedef struct _XawCache {
-  long value;
+  intptr_t value;
   XtPointer *elems;
   unsigned int num_elems;
 } XawCache;
@@ -395,7 +395,7 @@ qcmp_x_cache(register _Xconst void *left, register _Xconst void *right)
 static int
 bcmp_x_cache(register _Xconst void *pixmap, register _Xconst void *xaw)
 {
-  return (int)((long)pixmap - (long)((*(XawPixmap **)xaw)->pixmap));
+  return (int)((long)(uintptr_t)pixmap - (long)((*(XawPixmap **)xaw)->pixmap));
 }
 
 static int
@@ -415,7 +415,7 @@ qcmp_string(register _Xconst void *left, register _Xconst void *right)
 static int
 bcmp_long(register _Xconst void *value, register _Xconst void *cache)
 {
-  return (int)((long)value - (long)((*(XawCache **)cache)->value));
+  return (int)((intptr_t)value - (intptr_t)((*(XawCache **)cache)->value));
 }
 
 static int
@@ -448,7 +448,7 @@ _XawFindCache(XawCache *xaw,
     return (*cache);
 
   /* Colormap */
-  cache = (XawCache **)bsearch((void *)colormap, (*cache)->elems,
+  cache = (XawCache **)bsearch((void *)(uintptr_t)colormap, (*cache)->elems,
 			       (*cache)->num_elems, sizeof(XtPointer),
 			       bcmp_long);
   if (!cache || !(*cache)->num_elems)
@@ -457,7 +457,7 @@ _XawFindCache(XawCache *xaw,
     return (*cache);
 
   /* Depth */
-  cache = (XawCache **)bsearch((void *)(long)depth, (*cache)->elems,
+  cache = (XawCache **)bsearch((void *)(uintptr_t)depth, (*cache)->elems,
 			       (*cache)->num_elems, sizeof(XtPointer),
 			       bcmp_long);
 
@@ -493,7 +493,7 @@ _XawGetCache(XawCache *xaw, Screen *screen, Colormap colormap, int depth)
 		XtRealloc((char *)xaw->elems,
 			  (Cardinal)(sizeof(XtPointer) * (size_t)xaw->num_elems));
 	    }
-	  pcache->value = (long)screen;
+	  pcache->value = (intptr_t)screen;
 	  pcache->elems = NULL;
 	  pcache->num_elems = 0;
 	  xaw->elems[xaw->num_elems - 1] = (XtPointer)pcache;
@@ -746,7 +746,11 @@ BitmapLoader(XawParams *params, Screen *screen, Colormap colormap, int depth,
   if (params->name[0] != '/' && params->name[0] != '.')
     {
       if (!sub[0].substitution)
+	#ifdef _MSC_VER
+	sub[0].substitution = ".";
+	#else
 	sub[0].substitution = getenv("HOME");
+	#endif
       sub[1].substitution = (_XtString)params->name;
       if (pixmap_path == NULL)
 	GetResourcePixmapPath(DisplayOfScreen(screen));
diff --git a/src/Private.h b/src/Private.h
index 9067204..dd5f4d3 100644
--- a/src/Private.h
+++ b/src/Private.h
@@ -28,6 +28,8 @@
 #ifndef _XawPrivate_h
 #define _XawPrivate_h
 
+#include <unistd.h>
+
 #define XawMax(a, b) ((a) > (b) ? (a) : (b))
 #define XawMin(a, b) ((a) < (b) ? (a) : (b))
 #define XawAbs(a)    ((a) < 0 ? -(a) : (a))
diff --git a/src/Scrollbar.c b/src/Scrollbar.c
index ac75083..58bd653 100644
--- a/src/Scrollbar.c
+++ b/src/Scrollbar.c
@@ -795,7 +795,7 @@ NotifyScroll(Widget gw, XEvent *event, String *params, Cardinal *num_params)
 	    /*FALLTHROUGH*/
 	case 'F':
 	case 'f':
-	    XtCallCallbacks(gw, XtNscrollProc, (XtPointer)call_data);
+	    XtCallCallbacks(gw, XtNscrollProc, (XtPointer)(uintptr_t)call_data);
 	    break;
 	case 'C':
 	case 'c':    /* NotifyThumb has already called the thumbProc(s) */
diff --git a/src/Text.c b/src/Text.c
index 960214c..f06a9d6 100644
--- a/src/Text.c
+++ b/src/Text.c
@@ -1740,7 +1740,7 @@ static void
 HScroll(Widget w _X_UNUSED, XtPointer closure, XtPointer callData)
 {
     TextWidget ctx = (TextWidget)closure;
-    long pixels = (long)callData;
+    long pixels = (long)(uintptr_t)callData;
 
     if (pixels > 0) {
 	long max;
@@ -1770,7 +1770,7 @@ HJump(Widget w, XtPointer closure, XtPointer callData)
     pixels = ctx->text.left_margin -
 	     (ctx->text.r_margin.left - (int)(percent * (float)value));
 
-    HScroll(w, (XtPointer)ctx, (XtPointer)pixels);
+    HScroll(w, (XtPointer)ctx, (XtPointer)(uintptr_t)pixels);
 }
 
 /*
@@ -1833,7 +1833,7 @@ static void
 VScroll(Widget w _X_UNUSED, XtPointer closure, XtPointer callData)
 {
     TextWidget ctx = (TextWidget)closure;
-    long height, lines = (long)callData;
+    long height, lines = (long)(uintptr_t)callData;
 
     height = XtHeight(ctx) - VMargins(ctx);
     if (height < 1)
diff --git a/src/TextAction.c b/src/TextAction.c
index 09e22f2..6fe86b2 100644
--- a/src/TextAction.c
+++ b/src/TextAction.c
@@ -37,6 +37,7 @@ in this Software without prior written authorization from The Open Group.
 #include <X11/Xatom.h>
 #include <X11/Xfuncs.h>
 #include <X11/Xutil.h>
+#include <X11/Xwinsock.h>
 #include <X11/Xmu/Atoms.h>
 #include <X11/Xmu/Misc.h>
 #include <X11/Xmu/StdSel.h>
@@ -65,6 +66,9 @@ in this Software without prior written authorization from The Open Group.
 
 #define XawTextActionMaxHexChars	100
 
+#ifdef _MSC_VER
+#define usleep(time) Sleep(time)
+#endif
 /*
  * Prototypes
  */
diff --git a/src/TextPop.c b/src/TextPop.c
index 499c340..76cfda1 100644
--- a/src/TextPop.c
+++ b/src/TextPop.c
@@ -872,7 +872,7 @@ DoSearch(struct SearchAndReplace *search)
 #endif /* OLDXAW */
     }
 
-    dir = (XawTextScanDirection)(unsigned long)
+    dir = (XawTextScanDirection)(uintptr_t)
       ((XPointer)XawToggleGetCurrent(search->left_toggle) - R_OFFSET);
 
     pos = XawTextSearch(tw, dir, &text);
@@ -1027,7 +1027,7 @@ Replace(struct SearchAndReplace *search, Bool once_only, Bool show_current)
     else
 	replace.length = (int)strlen(replace.ptr);
 
-    dir = (XawTextScanDirection)(unsigned long)
+    dir = (XawTextScanDirection)(uintptr_t)
       ((XPointer)XawToggleGetCurrent(search->left_toggle) - R_OFFSET);
 
     redisplay = !once_only || (once_only && !show_current);
diff --git a/src/TextSink.c b/src/TextSink.c
index 8c85492..5625421 100644
--- a/src/TextSink.c
+++ b/src/TextSink.c
@@ -113,7 +113,7 @@ static XtResource resources[] = {
   },
 #ifndef OLDXAW
   {
-    XtNcursorColor,
+    (String)(uintptr_t)XtNcursorColor,
     XtCColor,
     XtRPixel,
     sizeof(Pixel),
@@ -1001,7 +1001,7 @@ static Cardinal num_prop_lists;
 static int
 bcmp_qident(_Xconst void *left, _Xconst void *right)
 {
-    return (int)((long)left - (*(XawTextProperty**)right)->identifier);
+    return (int)((long)(uintptr_t)left - (*(XawTextProperty**)right)->identifier);
 }
 
 static int
@@ -1149,7 +1149,7 @@ _XawTextSinkGetProperty(XawTextPropertyList *list, XrmQuark property)
 {
     if (property != NULLQUARK && list && list->properties) {
 	XawTextProperty **ptr = (XawTextProperty**)
-	    bsearch((void*)(long)property,
+	    bsearch((void*)(uintptr_t)property,
 		    list->properties, list->num_properties,
 		    sizeof(XawTextProperty*), bcmp_qident);
 
@@ -1492,7 +1492,7 @@ XawTextSinkConvertPropertyList(String name, String spec, Screen *screen,
     char buffer[BUFSIZ];
 
     if (prop_lists) ptr = (XawTextPropertyList**)
-	bsearch((void*)(long)qname, prop_lists, num_prop_lists,
+	bsearch((void*)(uintptr_t)qname, prop_lists, num_prop_lists,
 		sizeof(XawTextPropertyList*), bcmp_qident);
     if (ptr) {
 	propl = *ptr;
@@ -1745,7 +1745,7 @@ CvtStringToPropertyList(Display *dpy, XrmValue *args, Cardinal *num_args,
     if (w) {
 	XawTextPropertyList **ptr = NULL;
 	if (prop_lists) ptr = (XawTextPropertyList**)
-	    bsearch((void*)(long)XrmStringToQuark(name),
+	    bsearch((void*)(uintptr_t)XrmStringToQuark(name),
 		    prop_lists, num_prop_lists,
 		    sizeof(XawTextPropertyList*), bcmp_qident);
 
diff --git a/src/TextSrc.c b/src/TextSrc.c
index 0138e2a..9e7d5d6 100644
--- a/src/TextSrc.c
+++ b/src/TextSrc.c
@@ -1199,7 +1199,7 @@ exit_anchor_loop:
 	TellSourceChanged(src, left, right, block, lines);
 	/* Call callbacks, we have changed the buffer */
 	XtCallCallbacks(w, XtNcallback,
-			(XtPointer)((long)src->textSrc.changed));
+			(XtPointer)((uintptr_t)src->textSrc.changed));
     }
 
 #endif /* OLDXAW */
diff --git a/src/Toggle.c b/src/Toggle.c
index 2810890..052e96f 100644
--- a/src/Toggle.c
+++ b/src/Toggle.c
@@ -287,7 +287,7 @@ static void
 Notify(Widget w, XEvent *event _X_UNUSED, String *params _X_UNUSED, Cardinal *num_params _X_UNUSED)
 {
     ToggleWidget tw = (ToggleWidget)w;
-    long antilint = tw->command.set;
+    uintptr_t antilint = tw->command.set;
 
     XtCallCallbacks(w, XtNcallback, (XtPointer)antilint);
 }
diff --git a/src/Vendor.c b/src/Vendor.c
index 07faaa9..f8ede8b 100644
--- a/src/Vendor.c
+++ b/src/Vendor.c
@@ -111,7 +111,7 @@ static XtResource resources[] = {
  *
  ***************************************************************************/
 
-#if defined(__CYGWIN__) || defined(__MINGW32__) || defined(__APPLE__)
+#if defined(__CYGWIN__) || defined(__MINGW32__) || defined(__APPLE__) && !defined(_MSC_VER)
 /* to fix the EditRes problem because of wrong linker semantics */
 extern WidgetClass vendorShellWidgetClass; /* from Xt/Vendor.c */
 extern VendorShellClassRec _XawVendorShellClassRec;
diff --git a/src/Viewport.c b/src/Viewport.c
index 885e48d..6cad7d1 100644
--- a/src/Viewport.c
+++ b/src/Viewport.c
@@ -807,7 +807,7 @@ ScrollUpDownProc(Widget widget, XtPointer closure, XtPointer call_data)
 {
     ViewportWidget w = (ViewportWidget)closure;
     Widget child = w->viewport.child;
-    int pix = (int)(long)call_data;
+    int pix = (int)(uintptr_t)call_data;
     int x, y;
 
     if (child == NULL)
diff --git a/src/XawI18n.h b/src/XawI18n.h
index 7d20bfc..86f2243 100644
--- a/src/XawI18n.h
+++ b/src/XawI18n.h
@@ -39,6 +39,7 @@ in this Software without prior written authorization from The Open Group.
 #endif
 
 #include <stdlib.h>
+#include <unistd.h>
 
 #ifdef USE_XWCHAR_STRING
 int _Xwcslen
diff --git a/src/XawIm.c b/src/XawIm.c
index 5446c62..beead55 100644
--- a/src/XawIm.c
+++ b/src/XawIm.c
@@ -208,7 +208,7 @@ SetExtPart(VendorShellWidget w, XawVendorShellExtWidget vew)
     contextData = XtNew(contextDataRec);
     contextData->parent = (Widget)w;
     contextData->ve = (Widget)vew;
-    if (XSaveContext(XtDisplay(w), (Window)w, extContext, (char *)contextData)) {
+    if (XSaveContext(XtDisplay(w), (Window)(((intptr_t)w)&0xffffffff), extContext, (char *)contextData)) {
 	return(NULL);
     }
     return(&(vew->vendor_ext));
@@ -220,7 +220,7 @@ GetExtPart(VendorShellWidget w)
     contextDataRec *contextData;
     XawVendorShellExtWidget vew;
 
-    if (XFindContext(XtDisplay(w), (Window)w, extContext,
+	if (XFindContext(XtDisplay(w), (Window)(((intptr_t)w)& 0xffffffff), extContext,
 		      (XPointer*)&contextData)) {
 	return(NULL);
     }
@@ -320,7 +320,7 @@ static Widget SetErrCnxt(Widget w, XIM xim)
     contextErrData = XtNew(contextErrDataRec);
     contextErrData->widget = w;
     contextErrData->xim = xim;
-    if (XSaveContext(XtDisplay(w), (Window)xim, errContext,
+    if (XSaveContext(XtDisplay(w), (Window)(uintptr_t)xim, errContext,
 	(char *)contextErrData)) {
 	return(NULL);
     }
@@ -376,11 +376,11 @@ DestroyAllIM(XawVendorShellExtPart *ve)
     /*
      * Close Input Method
      */
-    if (!XFindContext(XDisplayOfIM(ve->im.xim), (Window)ve->im.xim, errContext,
+    if (!XFindContext(XDisplayOfIM(ve->im.xim), (Window)(uintptr_t)ve->im.xim, errContext,
 		      (XPointer*)&contextErrData)) {
 	if (contextErrData) XtFree((char *)contextErrData);
     }
-    XDeleteContext(XDisplayOfIM(ve->im.xim), (Window)ve->im.xim, errContext);
+    XDeleteContext(XDisplayOfIM(ve->im.xim), (Window)(uintptr_t)ve->im.xim, errContext);
     CloseIM(ve);
     ve->im.xim = NULL;
 
@@ -400,11 +400,11 @@ FreeAllDataOfVendorShell(XawVendorShellExtPart *ve, VendorShellWidget vw)
     XawIcTableList       p, next;
     contextErrDataRec *contextErrData;
 
-    if (!XFindContext(XtDisplay(vw), (Window)vw, extContext,
+    if (!XFindContext(XtDisplay(vw), (Window)(uintptr_t)vw, extContext,
 		      (XPointer*)&contextErrData)) {
 	if (contextErrData) XtFree((char *)contextErrData);
     }
-    XDeleteContext(XtDisplay(vw), (Window)vw, extContext);
+    XDeleteContext(XtDisplay(vw), (Window)(uintptr_t)vw, extContext);
     if (ve->ic.shared_ic_table)
         XtFree((char *)ve->ic.shared_ic_table);
     if (ve->im.resources) XtFree((char *)ve->im.resources);
@@ -800,27 +800,27 @@ CreateIC(Widget w, XawVendorShellExtPart *ve)
 	}
 	if (p->flg & CIFg) {
 	    ADD_PE(XNForeground);
-	    ADD_PE(p->foreground);
+	    ADD_PE((uintptr_t)p->foreground);
 	    ADD_ST(XNForeground);
-	    ADD_ST(p->foreground);
+	    ADD_ST((uintptr_t)p->foreground);
 	}
 	if (p->flg & CIBg) {
 	    ADD_PE(XNBackground);
-	    ADD_PE(p->background);
+	    ADD_PE((uintptr_t)p->background);
 	    ADD_ST(XNBackground);
-	    ADD_ST(p->background);
+	    ADD_ST((uintptr_t)p->background);
 	}
 	if (p->flg & CIBgPixmap) {
 	    ADD_PE(XNBackgroundPixmap);
-	    ADD_PE(p->bg_pixmap);
+	    ADD_PE((uintptr_t)p->bg_pixmap);
 	    ADD_ST(XNBackgroundPixmap);
-	    ADD_ST(p->bg_pixmap);
+	    ADD_ST((uintptr_t)p->bg_pixmap);
 	}
 	if (p->flg & CILineS) {
 	    ADD_PE(XNLineSpace);
-	    ADD_PE(p->line_spacing);
+	    ADD_PE((uintptr_t)p->line_spacing);
 	    ADD_ST(XNLineSpace);
-	    ADD_ST(p->line_spacing);
+	    ADD_ST((uintptr_t)p->line_spacing);
 	}
     }
 
@@ -865,11 +865,11 @@ CreateIC(Widget w, XawVendorShellExtPart *ve)
     }
 
     ADD_IC(XNInputStyle);
-    ADD_IC(p->input_style);
+    ADD_IC((uintptr_t)p->input_style);
     ADD_IC(XNClientWindow);
-    ADD_IC(XtWindow(ve->parent));
+    ADD_IC((uintptr_t)XtWindow(ve->parent));
     ADD_IC(XNFocusWindow);
-    ADD_IC(XtWindow(w));
+    ADD_IC((uintptr_t)XtWindow(w));
 
     if (pe_cnt > 0) {
 	END_PE();
@@ -962,27 +962,27 @@ SetICValues(Widget w, XawVendorShellExtPart *ve, Bool focus)
 	}
 	if (p->flg & CIFg) {
 	    ADD_PE(XNForeground);
-	    ADD_PE(p->foreground);
+	    ADD_PE((uintptr_t)p->foreground);
 	    ADD_ST(XNForeground);
-	    ADD_ST(p->foreground);
+	    ADD_ST(uintptr_t)(p->foreground);
 	}
 	if (p->flg & CIBg) {
 	    ADD_PE(XNBackground);
-	    ADD_PE(p->background);
+	    ADD_PE((uintptr_t)p->background);
 	    ADD_ST(XNBackground);
-	    ADD_ST(p->background);
+	    ADD_ST((uintptr_tp->background);
 	}
 	if (p->flg & CIBgPixmap) {
 	    ADD_PE(XNBackgroundPixmap);
-	    ADD_PE(p->bg_pixmap);
+	    ADD_PE((uintptr_t)p->bg_pixmap);
 	    ADD_ST(XNBackgroundPixmap);
-	    ADD_ST(p->bg_pixmap);
+	    ADD_ST((uintptr_t)p->bg_pixmap);
 	}
 	if (p->flg & CILineS) {
 	    ADD_PE(XNLineSpace);
-	    ADD_PE(p->line_spacing);
+	    ADD_PE((uintptr_t)p->line_spacing);
 	    ADD_ST(XNLineSpace);
-	    ADD_ST(p->line_spacing);
+	    ADD_ST((uintptr_t)p->line_spacing);
 	}
     }
     if (p->input_style & XIMPreeditPosition) {
@@ -1029,7 +1029,7 @@ SetICValues(Widget w, XawVendorShellExtPart *ve, Bool focus)
     }
     if (focus == TRUE) {
 	ADD_IC(XNFocusWindow);
-	ADD_IC(XtWindow(w));
+	ADD_IC((uintptr_t)XtWindow(w));
     }
     if (ic_cnt > 0) {
 	END_IC();
@@ -1395,12 +1395,12 @@ Destroy(Widget w, XawVendorShellExtPart *ve)
     XtFree( (char*) ve->im.resources );
 
     if (extContext != (XContext)0 &&
-	!XFindContext (XtDisplay (w), (Window)w,
+	!XFindContext (XtDisplay (w), (Window)(uintptr_t)w,
 		       extContext, (XPointer*)&contextData))
         XtFree( (char*) contextData );
 
     if (errContext != (XContext)0 &&
-	!XFindContext (XDisplayOfIM( ve->im.xim ), (Window) ve->im.xim,
+	!XFindContext (XDisplayOfIM( ve->im.xim ), (Window)(uintptr_t) ve->im.xim,
 		       errContext, (XPointer*) &contextErrData))
         XtFree( (char*) contextErrData );
 }
